### PR TITLE
Update preliminary project structure and build files

### PR DESCRIPTION
**Pull request by leoliu49**
Prof. Dietl suggested I break up PR https://github.com/opprop/universe/pull/3 into smaller chunks.
For this part, I cherrypicked commits so that the build and configuration files are set appropriately.

Next steps:
1. Make a separate PR for project file organization refactoring
2. Apply remainder of the commits in https://github.com/opprop/universe/pull/3

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,135 @@
+apply plugin: 'java'
+
+/************************ system variables ************************/
+def env         = System.getenv()
+
+def javaHome    = env["JAVA_HOME"]
+
+//TODO: copy from CF Inference, need refactor later on
+//determined based on the the version of the currently running JVM
+def jreVersion  = {
```
Comment (wmdietl) [inline build.gradle:10]: @xingweitian @leoliu49 Can you have a look through the build.gradle and make it consistent with the current CFI build style?

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+dependencies {
+    // solver backend dependencies
+    // TODO: should be removed when we get rid of classic solver
+    // TODO: currently the runtime dependencies of sat4j is bit tricky --- it actually uses the sat4j from cfi/dist
+    //       should we also snapshot a copy of sat4j from mavenCentral into ontology?
+    compile 'org.ow2.sat4j:org.ow2.sat4j.core:2.3.4'
+    compile 'org.ow2.sat4j:org.ow2.sat4j.maxsat:2.3.4'
+    compile group: 'junit', name: 'junit', version: '4.12'
```
Comment (wmdietl) [inline build.gradle:77]: I would use one consistent style, so either the `x:y:z` style or the `group: x, name: y, version: z` style.

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+//Switch the Javac used to the JSR308
+tasks.compileJava {
+  description = 'Compiles this project using the jsr308 compiler.'
+  options.fork = true
+  options.compilerArgs = [
+      '-implicit:class',
+      '-Awarns', '-Xmaxwarns', '10000']
+  if(jreVersion == 1.7) {
```
Comment (wmdietl) [inline build.gradle:117]: We no longer support java 7, so special casing that can be removed.
Instead, we need to special case Java 8 and use the error-prone compiler if necessary.

File: scripts/env-setup.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
```
Comment (wmdietl) [inline scripts/env-setup.sh:1]: Could you also add a top-level README.md file that gives an overview of all the different scripts and when to use them?

File: scripts/env-setup.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+
+export JAVA_HOME=${JAVA_HOME:-$(dirname $(dirname $(dirname $(readlink -f $(/usr/bin/which java)))))}
+
+export JSR308=$(cd $(dirname "$0")/../.. && pwd)
```
Comment (wmdietl) [inline scripts/env-setup.sh:5]: Make sure that no file has the "No newline at EOF" warning sign.

File: scripts/projects-setup.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+##### Clone checker-framework-inference
+if [ -d $JSR308/checker-framework-inference ] ; then
+    (cd $JSR308/checker-framework-inference && git pull)
+else
+    (cd $JSR308 && git clone -b hackathon --depth 1 https://github.com/"$REPO_SITE"/checker-framework-inference.git)
+fi
+
+##### Clone generic-type-inference-solver
+if [ -d $JSR308/generic-type-inference-solver ] ; then
```
Comment (wmdietl) [inline scripts/projects-setup.sh:39]: This project is dead and we shouldn't have to clone it.

File: scripts/projects-setup.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    (cd $JSR308/annotation-tools && git pull)
+else
+    (cd $JSR308 && git clone --depth 1 https://github.com/"$REPO_SITE"/annotation-tools.git)
+fi
+
+##### Clone checker-framework-inference
+if [ -d $JSR308/checker-framework-inference ] ; then
+    (cd $JSR308/checker-framework-inference && git pull)
+else
+    (cd $JSR308 && git clone -b hackathon --depth 1 https://github.com/"$REPO_SITE"/checker-framework-inference.git)
```
Comment (wmdietl) [inline scripts/projects-setup.sh:35]: Is it enough to clone this project and then call `gradlew cloneAndBuildDependencies` in that project?
https://github.com/opprop/checker-framework-inference/blob/master/build.gradle#L249

File: scripts/projects-setup.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    (cd $JSR308/generic-type-inference-solver && git pull)
+else
+    (cd $JSR308 && git clone --depth 1 https://github.com/"$REPO_SITE"/generic-type-inference-solver.git)
+fi
+
+# Build jsr308-langtools
+(cd $JSR308/annotation-tools/ && ./.travis-build-without-test.sh)
+
+## Build checker-framework, with jdk
+ant -f $JSR308/checker-framework/checker/build.xml jar
```
Comment (wmdietl) [inline scripts/projects-setup.sh:49]: These build commands also look out-of-date.

File: src/GUT/GUTAnnotatedTypeFactory.java [hunk 9]
```diff
@@ -91,6 +71,11 @@ public GUTAnnotatedTypeFactory(BaseTypeChecker checker) {
         this.postInit();
     }
 
+    @Override
+    public AnnotatedTypeMirror getAnnotatedType(Element elt) {
```
Comment (wmdietl) [inline src/GUT/GUTAnnotatedTypeFactory.java:75]: There is no point in overriding a method and then only calling super.

File: src/GUT/GUTVisitor.java [hunk 10]
```diff
@@ -16,31 +28,17 @@
...
-import com.sun.source.tree.Tree;
-import com.sun.source.tree.Tree.Kind;
-import com.sun.source.tree.TypeCastTree;
-import com.sun.source.tree.VariableTree;
+import java.util.List;
+import java.util.Set;
 
 /**
- * Type visitor to enforce the GUT type rules.
+ * Type visitor to enforce the gut type rules.
```
Comment (wmdietl) [inline src/GUT/GUTVisitor.java:41]: The upper-case `GUT` seems easier to read to me.

File: src/GUT/GUTVisitor.java [hunk 11]
```diff
@@ -99,7 +97,7 @@ public Void visitMethod(MethodTree node, Void p) {
         if (node.getReceiverParameter() != null &&
                 !node.getReceiverParameter().getModifiers().getAnnotations()
                         .isEmpty()) {
-                    checker.report(Result.failure("uts.receiver.annotations.forbidden"),node);
+                    //checker.report(Result.failure("uts.receiver.annotations.forbidden"),node);
```
Comment (wmdietl) [inline src/GUT/GUTVisitor.java:100]: There at least needs to be a TODO here why the error is disabled.

File: src/GUTI/GUTIAnnotatedTypeFactory.java [hunk 12]
```diff
@@ -29,20 +43,65 @@ public GUTIAnnotatedTypeFactory(InferenceChecker inferenceChecker,
...
+        }
+        return atm;
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(new ImplicitsTreeAnnotator(this), new GUTIInferenceTreeAnnotater(this,
+                realChecker, realTypeFactory, new GUTIVariableAnnotator(this, realTypeFactory, realChecker, slotManager, constraintManager), slotManager));
+    }
+/*    @Override
```
Comment (wmdietl) [inline src/GUTI/GUTIAnnotatedTypeFactory.java:63]: Let's not add debugging artifacts like this one.

File: src/GUTI/GUTIAnnotatedTypeFactory.java [hunk 13]
```diff
@@ -29,20 +43,65 @@ public GUTIAnnotatedTypeFactory(InferenceChecker inferenceChecker,
...
      * The type of "this" is always "self".
      */
     @Override
     public AnnotatedDeclaredType getSelfType(Tree tree) {
         AnnotatedDeclaredType type = super.getSelfType(tree);
-        if (type != null) {
-            GUTAnnotatedTypeFactory gutATF = (GUTAnnotatedTypeFactory) InferenceMain
-                    .getInstance().getRealTypeFactory();
-            type.replaceAnnotation(gutATF.SELF);
+        GUTAnnotatedTypeFactory gutATF = (GUTAnnotatedTypeFactory) realTypeFactory;
```
Comment (wmdietl) [inline src/GUTI/GUTIAnnotatedTypeFactory.java:75]: The try-block here is really ugly. Can you see whether this actually still crashes?
Add a big TODO that this should be cleaned up.
I'm not sure whether we need to create an equality constraint here. Maybe the old implementation was good enough?

File: src/GUTI/GUTIAnnotatedTypeFactory.java [hunk 14]
```diff
@@ -52,6 +111,37 @@ public VariableAnnotator createVariableAnnotator(
...
+        }
+
+        @Override
+        public Void visitTypeCast(TypeCastTree typeCast, AnnotatedTypeMirror atm) {
+            super.visitTypeCast(typeCast, atm);
+            if (!(atm instanceof AnnotatedPrimitiveType)) {
+                GUTAnnotatedTypeFactory gutATF = (GUTAnnotatedTypeFactory) InferenceMain.getInstance().getRealTypeFactory();
+                SlotManager sm = InferenceMain.getInstance().getSlotManager();
+                ConstraintManager cm = InferenceMain.getInstance().getConstraintManager();
+                cm.addInequalityConstraint(sm.getVariableSlot(atm), sm.getSlot(gutATF.BOTTOM));
```
Comment (wmdietl) [inline src/GUTI/GUTIAnnotatedTypeFactory.java:135]: I would also put `sm.getVariableSlot(atm)` into a local variable.

File: src/GUTI/GUTIAnnotatedTypeFactory.java [hunk 15]
```diff
@@ -52,6 +111,37 @@ public VariableAnnotator createVariableAnnotator(
...
+                cm.addInequalityConstraint(sm.getVariableSlot(atm), sm.getSlot(gutATF.LOST));
+                cm.addInequalityConstraint(sm.getVariableSlot(atm), sm.getSlot(gutATF.VPLOST));
+                cm.addInequalityConstraint(sm.getVariableSlot(atm), sm.getSlot(gutATF.ANY));
```
Comment (wmdietl) [inline src/GUTI/GUTIAnnotatedTypeFactory.java:138]: I'm not sure why ANY should be forbidden for a cast type.

File: src/GUTI/GUTIChecker.java [hunk 16]
```diff
@@ -58,4 +57,10 @@ public InferenceAnnotatedTypeFactory createInferenceATF(
                 slotManager, constraintManager);
     }
 
+    @Override
+    public boolean isInsertMainModOfLocalVar() {
+        return true;
```
Comment (wmdietl) [inline src/GUTI/GUTIChecker.java:62]: I think we don't want this - local variables should be top and then refined automatically.

File: src/GUTI/GUTIMaxSatSerializer.java [hunk 17]
```diff
@@ -71,79 +71,91 @@ public GUTIMaxSatSerializer(Lattice lattice) {
                             (VariableSlot) decl, (VariableSlot) result,
                             constraint);
                 }
+
+                /*for (int i=0; i<resultClauses.length; i++) {
```
Comment (wmdietl) [inline src/GUTI/GUTIMaxSatSerializer.java:75]: Don't add debugging code like this.

File: src/GUTI/GUTIVisitor.java [hunk 18]
```diff
@@ -1,67 +1,78 @@
...
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.DeclaredType;
+import java.util.List;
 
 /**
- * Type visitor to either enforce or infer the GUT type rules.
+ * Type visitor to either enforce or infer the gut type rules.
```
Comment (wmdietl) [inline src/GUTI/GUTIVisitor.java:61]: Throughout these changes, don't change `GUT` to `gut`.

File: src/GUTI/GUTIVisitor.java [hunk 19]
```diff
@@ -98,35 +115,71 @@ protected GUTIInferenceValidator createTypeValidator() {
         /*
         GUTAnnotatedTypeFactory gutATFParameter = (GUTAnnotatedTypeFactory) InferenceMain.getInstance().getRealTypeFactory();
         return new GUTIInferenceValidator(checker, this, gutATFParameter);*/
-        return new GUTIInferenceValidator(checker, this, this.atypeFactory);
+        //return new GUTIInferenceValidator(checker, this, atypeFactory);
```
Comment (wmdietl) [inline src/GUTI/GUTIVisitor.java:118]: I don't see a point in this change.

File: tests/GUTI/TypeVariableUse.java [hunk 20]
```diff
@@ -1,11 +0,0 @@
-import GUT.qual.*;
```
Comment (wmdietl) [inline tests/GUTI/TypeVariableUse.java:1]: Why do a few changes appear as removed-adds instead of as moves?

File: tst/GUT/GUTTestsLostNo.java [hunk 21]
```diff
@@ -0,0 +1,22 @@
+package GUT;
+
+import java.io.File;
```
Comment (wmdietl) [inline tst/GUT/GUTTestsLostNo.java:3]: Is `tst` the best name for this directory?

**Review by wmdietl — CHANGES_REQUESTED** [review]
@leoliu49 Thanks for working on breaking down this large PR! I have lots of small comments that would be great to look into.
Maybe it would make sense to further break this down into smaller PRs:
- remove Eclipse files,
- update version control files,
- move test files,
- etc...

That will make the PRs easier to review and the history cleaner.

File: build.gradle (hunk 1 above)
Comment (xingweitian) [reply build.gradle:10]: Just found this comment...
I made the change in another branch before: https://github.com/opprop/universe/blob/adapt-to-cf-master/build.gradle

Comment (xingweitian) [reply build.gradle:10]: This PR should merge with the branch `adapt-to-cf-master`. We used to work on that one.

**Pull request closed without merging**